backend/asm_targets/asm_directives.ml

```diff
@@ -0,0 +1,322 @@
+        { section = section2; str = str2; comment = comment2 } =
+      let c = Asm_section.compare section1 section2 in
+      if c <> 0
+      then c
+      else
+        let c = String.compare str1 str2 in
+        if c <> 0 then c else Option.compare String.compare comment1 comment2
+
+    let equal t1 t2 = compare t1 t2 = 0
+
+    let hash t = Hashtbl.hash t
+
+    let print _ _ = Misc.fatal_error "Not yet implemented"
+
+    let output _ _ = Misc.fatal_error "Not yet implemented"
+  end)
+end
+
+module Make (A : Asm_directives_intf.Arg) : Asm_directives_intf.S = struct
+  module Int8 = Numbers.Int8
+  module Int16 = Numbers.Int16
+  module Uint8 = Numbers.Uint8
+  module Uint16 = Numbers.Uint16
+  module Uint32 = Numbers.Uint32
+  module Uint64 = Numbers.Uint64
+  module D = A.D
+
+  let sections_seen = ref []
+
+  let current_section_ref = ref None
+
+  let cached_strings = ref Cached_string.Map.empty
+
+  let temp_var_counter = ref 0
+
+  let is_macos () =
+    match Target_system.assembler () with
+    | MASM | GAS_like -> false
+    | MacOS -> true
+
+  let if_not_masm f =
+    match Target_system.assembler () with
+    | MASM -> ()
+    | GAS_like | MacOS -> f ()
+
+  let loc ~file_num ~line ~col =
+    if_not_masm (fun () -> D.loc ~file_num ~line ~col ())
+
+  let new_line () = D.new_line ()
+
+  let not_initialized () =
+    Misc.fatal_error "[Asm_directives.initialize] has not been called"
+
+  let define_label label =
+    let lbl_section = Asm_label.section label in
+    let this_section =
+      match !current_section_ref with
+      | None -> not_initialized ()
+      | Some this_section -> this_section
+    in
+    if not (Asm_section.equal lbl_section this_section)
+    then
+      Misc.fatal_errorf
+        "Cannot define label %a intended for section %a in section %a"
+        Asm_label.print label Asm_section.print lbl_section Asm_section.print
+        this_section;
+    (* CR-someday bkhajwal: Be aware of MASM label type annotation *)
+    D.label (Asm_label.encode label)
+
+  let switch_to_section section =
+    let first_occurrence =
+      if List.mem section !sections_seen
+      then false
+      else begin
+        sections_seen := section :: !sections_seen;
+        true
+      end
+    in
+    match !current_section_ref with
+    | Some section' when Asm_section.equal section section' ->
+      assert (not first_occurrence);
+      ()
+    | _ ->
+      current_section_ref := Some section;
+      let ({ names; flags; args } : Asm_section.flags_for_section) =
+        Asm_section.flags section ~first_occurrence
+      in
+      if not first_occurrence then new_line ();
+      D.section names flags args;
+      if first_occurrence then define_label (Asm_label.for_section section)
+
+  let initialize () =
+    cached_strings := Cached_string.Map.empty;
+    sections_seen := [];
+    temp_var_counter := 0;
+    (* Forward label references are illegal in GAS *)
```

Comment:
`current_section_ref` isn't reset in `initialize`, is it intentional? I think it might cause the assertion on L124 to fail for GAS (call from L143). 
Also, it would be nice to rename `current_section_ref` to `current_dwarf_section_ref`, to avoid confusion with `D.text ()` in L151, until text and  data section representation are included in `Asm_section`.